test/src/utils/differentiator.test.cpp

```diff
@@ -0,0 +1,328 @@
+ *
+ *    Copyright 2018 HYPED
+ *    Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ *    except in compliance with the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software distributed under
+ *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ *    either express or implied. See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <iostream>
+#include "math.h"
+#include "gtest/gtest.h"
+#include "utils/math/differentiator.hpp"
+#include "utils/system.hpp"
+#include "data/data_point.hpp"
+
+using hyped::data::DataPoint;
+using hyped::utils::math::Differentiator;
+
+// -------------------------------------------------------------------------------------------------
+// Functionality
+// -------------------------------------------------------------------------------------------------
+
+/**
+ * Struct used for test fixtures testing functionality.
+ * Constructor sets the values of the DataPoint objects.
+ * Helper functions defined to assist in testing
+ */
+struct DifferentiatorFunctionality : public ::testing::Test {
+ protected:
+  Differentiator<float> diff_test;
+  DataPoint<float> test_point;
+  DataPoint<float> second_point;
+  DataPoint<float> third_point;
+
+  void SetUp()
+  {
+    // Sets values of Data points
+    test_point.value = 10.5;
+    test_point.timestamp = 1;
+
+    second_point.value = 18.6;
+    second_point.timestamp = 2;
+
+    third_point.value= second_point.value;
+    third_point.timestamp = 3;
+  }
+
+  /**
+   * @brief Returns the difference between two data point values.
+  */
+  float deltaValue(DataPoint<float> point_1, DataPoint<float> prev_point)
+  {
+    return point_1.value - prev_point.value;
+  }
+
+  /**
+   * @brief Returns the difference between two data point timestamps
+  */
+  float deltaTime(DataPoint<float> point_1, DataPoint<float> prev_point)
+  {
+    return point_1.timestamp- prev_point.timestamp;
+  }
+
+  /**
+   * @brief Returns the gradient between two data points
+  */
+  float gradientOfPoints(DataPoint<float> point_1, DataPoint<float> prev_point)
+  {
+    return (deltaValue(point_1, prev_point))/((deltaTime(point_1, prev_point))/1e6);
+  }
+  void TearDown() {}
+};
+
+/**
+ * Test fixture used for determining whether initialisation works correctly.
+ * The value should always return 0
+*/
+TEST_F(DifferentiatorFunctionality, differentiatorInitialisedValue)
+{
+  ASSERT_EQ(diff_test.update(test_point).value, 0)
+    << "Initalised value should be 0";
+}
+
+/**
+ * Test fixture used for determining whether initialisation works correctly.
+ * The timestamp should always be returned unchanged
+*/
+TEST_F(DifferentiatorFunctionality, differentiatorInitialisedTimestamp)
```

Comment:
As a general comment for future tests, the name of the test should start with a verb. for example. handlesInitialTimestep